docs/standard/native-interop/tutorial-comwrappers.md

```diff
@@ -0,0 +1,430 @@
+
+```csharp
+[UnmanagedCallersOnly]
+public static int GetString(IntPtr _this, IntPtr* str);
+```
+
+The wrapper implementation of `IDemoGetType.GetString()` should consist of marshalling logic and then a dispatch to the managed object being wrapped. All the state for dispatch is contained within the provided `_this` argument. The `_this` argument will actually be of type `ComInterfaceDispatch*`. This type represents a low-level structure with a single field, `Vtable`, that will be discussed later. Further details of this type and its layout are an implementation detail of the runtime and should not be depended upon. In order to retrieve the managed instance from a `ComInterfaceDispatch*` instance, use the following code:
+
+```csharp
+IDemoGetType inst = ComInterfaceDispatch.GetInstance<IDemoGetType>((ComInterfaceDispatch*)_this);
+```
+
+Now that you have a C# method that can be inserted into a vtable, you can construct the vtable. Note the use of [`RuntimeHelpers.AllocateTypeAssociatedMemory()`][api_allocatetypeassociatedmemory] for allocating memory in a way that works with [unloadable][doc_unloadability] assemblies.
+
+```csharp
+GetIUnknownImpl(
+    out IntPtr fpQueryInterface,
+    out IntPtr fpAddRef,
+    out IntPtr fpRelease);
+
+{
+    int tableCount = 4;
+    int idx = 0;
+    var vtable = (IntPtr*)RuntimeHelpers.AllocateTypeAssociatedMemory(
+        typeof(DemoComWrappers),
+        IntPtr.Size * tableCount);
+    vtable[idx++] = fpQueryInterface;
+    vtable[idx++] = fpAddRef;
+    vtable[idx++] = fpRelease;
+    vtable[idx++] = (IntPtr)(delegate* unmanaged<IntPtr, IntPtr*, int>)&ABI.
+IDemoGetTypeManagedWrapper.GetString;
+    Debug.Assert(tableCount == idx);
+    s_IDemoGetTypeVTable = (IntPtr)vtable;
+}
+```
+
+The allocation of vtables is the first part of implementing `ComputeVtables()`. You should also construct comprehensive definitions for types that you're planning to support &ndash; think `DemoImpl`. Using the constructed vtables, you can now create a series of [`ComInterfaceEntry`][api_cominterfaceentry] instances that represent the complete view of the managed object in COM.
+
+```csharp
+{
+    s_DemoImplDefinitionLen = 2;
+    int idx = 0;
+    var entries = (ComInterfaceEntry*)RuntimeHelpers.AllocateTypeAssociatedMemory(
+        typeof(DemoComWrappers),
+        sizeof(ComInterfaceEntry) * s_DemoImplDefinitionLen);
+    entries[idx].IID = IDemoGetType.IID_IDemoGetType;
+    entries[idx++].Vtable = s_IDemoGetTypeVTable;
+    entries[idx].IID = IDemoStoreType.IID_IDemoStoreType;
+    entries[idx++].Vtable = s_IDemoStoreVTable;
+    Debug.Assert(s_DemoImplDefinitionLen == idx);
+    s_DemoImplDefinition = entries;
+}
+```
```

Comment:
Comment on the full code in the samples repo:
```C#
Debug.Assert(flags.HasFlag(CreateComInterfaceFlags.None));
```

The condition is always true since `None == 0`. It should probably be `Debug.Assert(flags is CreateComInterfaceFlags.None)`.